Update skopeo readme and man page (#917)
containers/skopeo: README.md

`skopeo` is a command line utility that performs various operations on container images and image repositories.

> `skopeo` does not require to be running as root to do most of its operations.

Reviewer: require -> require the user

> `skopeo` does not require to be running as root to do most of its operations.

Reviewer: ```suggestion
`skopeo` does not require the user to be running as root to do most of its operations.
```

```

# skopeo copy example:
```console
$ skopeo copy --src-creds=testuser:testpassword docker://myregistrydomain.com:5000/private oci:local_oci_image
```
If your cli config is found but it doesn't contain the necessary credentials for the queried registry
you'll get an error. You can fix this by either logging in (via `docker login`) or providing `--creds` or `--src-creds|--dest-creds`.

Obtaining skopeo

> -

Reviewer 2: This `-` apparently makes the line above a header.

 Skopeo operates on the following image and repository types:

 * containers-storage:docker-reference

>          An image located in a local containers/storage image store.  Location and image store specified in /etc/containers/storage.conf  (Backend for [Podman](https://podman.io), [CRI-O](https://cri-o.io), [Buildah](https://buildah.io) and friends)

Reviewer: ```suggestion
         An image located in a local containers/storage image store.  Both the location and image store are specified in /etc/containers/storage.conf  (Backend for [Podman](https://podman.io), [CRI-O](https://cri-o.io), [Buildah](https://buildah.io) and friends)
```

>          An image located in a local containers/storage image store.  Both the location and image store are specified in /etc/containers/storage.conf  (Backend for [Podman](https://podman.io), [CRI-O](https://cri-o.io), [Buildah](https://buildah.io) and friends)

Reviewer 2: ```suggestion
         An image located in a local containers/storage image store.  Both the location and image store are specified in `/etc/containers/storage.conf`.  (This is the backend for [Podman](https://podman.io), [CRI-O](https://cri-o.io), [Buildah](https://buildah.io) and friends)
```

 * dir:path
         An existing local directory path storing the manifest, layer tarballs and signatures as individual files. This is a non-standardized format, primarily useful for debugging or noninvasive container inspection.

 * docker://docker-reference

>          An image in a registry implementing the "Docker Registry HTTP API V2". By default, uses the authorization state in either `$XDG_RUNTIME_DIR/containers/auth.json`, which is set using `(skopeo login)`.

Reviewer 2: “either” … missing “or”

 * docker-archive:path[:docker-reference]

>          An image is stored in the `podman save` or `docker save` formated file.  docker-reference is only used when creating such a file, and it must not contain a digest.

Reviewer 2: `podman save` supports multiple formats; “in a `podman save --format docker-archive` format” is pointless IMHO.

Also,
```suggestion
         An image is stored in a `docker save`-formatted file.  docker-reference is only used when creating such a file, and it must not contain a digest.
```

$ skopeo copy docker://busybox:latest dir:existingemptydirectory
$ skopeo copy docker://busybox:latest oci:busybox_ocilayout:latest
```console
$ skopeo copy docker://quay.io/buildah/stable docker://registry.internal.company.com/buildah
$ skopeo copy oci:busybox_ocilayout:latest dir:existingemptydirectory
```

Deleting images
-
For example,
```sh
```console
$ skopeo delete docker://localhost:5000/imagename:latest
```

Private registries with authentication
-
When interacting with private registries, `skopeo` first looks for `--creds` (for `skopeo inspect|delete`) or `--src-creds|--dest-creds` (for `skopeo copy`) flags. If those aren't provided, it looks for the Docker's cli config file (usually located at `$HOME/.docker/config.json`) to get the credentials needed to authenticate. The ultimate fallback, as Docker does, is to provide an empty authentication when interacting with those registries.
skopeo uses credentials from the --creds (for skopeo inspect|delete) or --src-creds|--dest-creds (for skopeo copy) flags, if set; otherwise it uses configuration set by skopeo login, podman login, buildah login, or docker login.

Examples:

> ```sh

Reviewer 2: This, or ```console, should remain; the example is now formatted as free-flowing text.

Author: fixed

> $ skopeo inspect --creds=testuser:testpassword docker://myregistrydomain.com:5000/busybox

Reviewer: need 3 backticks here too, otherwise more freeflowing text

Author: They are all fixed now.  PTAL

Inspecting a repository
-
`skopeo` is able to _inspect_ a repository on a container registry and fetch images layers.
The _inspect_ command fetches the repository's manifest and it is able to show you a `docker inspect`-like
json output about a whole repository or a tag. This tool, in contrast to `docker inspect`, helps you gather useful information about
a repository or a tag before pulling it (using disk space).  The inspect command can show you which tags are available for the given 
repository, the labels the image has, the creation date and operating system of the image and more.  

Examples:

> ```sh

Reviewer 2: `console` might look better than `sh`. (I haven’t tried it for this content.)

Author: Switched

# show properties of fedora:latest
$ skopeo inspect docker://registry.fedoraproject.org/fedora:latest
{
    "Name": "registry.fedoraproject.org/fedora",
    "Digest": "sha256:655721ff613ee766a4126cb5e0d5ae81598e1b0c3bcf7017c36c4d72cb092fe9",
    "RepoTags": [
        "24",
        "25",
        "26-modular",
        "26",
        "27-aarch64",
        "27-armhfp",
        "27-ppc64le",
        "27-x86_64",

>         "27",

Reviewer 2: Maybe replace most of the values with a `…` if we want the users to read on.

        "28-aarch64",
        "28-armhfp",
        "28-ppc64le",
        "28-x86_64",
        "28",
        "29-aarch64",
        "29-ppc64le",
        "29-s390x",
        "29-x86_64",
        "29",
        "30-aarch64",
        "30-ppc64le",
        "30-s390x",
        "30-x86_64",
        "30",
        "latest",
        "rawhide",
        "30-armhfp",
        "29-armhfp",
        "31-aarch64",
        "31-x86_64",
        "31",
        "31-armhfp",
        "31-s390x",
        "31-ppc64le",
        "32-aarch64",
        "32-ppc64le",
        "32-s390x",
        "32-x86_64",
        "32",
        "33-aarch64",
        "33-ppc64le",
        "33-s390x",
        "33-x86_64",
        "33"
    ],
    "Created": "2020-04-29T06:48:16Z",
    "DockerVersion": "1.10.1",
    "Labels": {
        "license": "MIT",
        "name": "fedora",
        "vendor": "Fedora Project",
        "version": "32"
    },
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:3088721d7dbf674fc0be64cd3cf00c25aab921cacf35fa0e7b1578500a3e1653"
    ],
    "Env": [
        "DISTTAG=f32container",
        "FGC=f32",
        "container=oci"
    ]
}

> # Show configuration data about fedora lates

Reviewer 2: ```suggestion
# Show container configuration from `fedora:latest`
```
or … from a latest Fedora image.

$ skopeo inspect --config docker://registry.fedoraproject.org/fedora:latest  | jq
{
  "created": "2020-04-29T06:48:16Z",
  "architecture": "amd64",
  "os": "linux",
  "config": {
    "Env": [
      "DISTTAG=f32container",
      "FGC=f32",
      "container=oci"
    ],
    "Cmd": [
      "/bin/bash"
    ],
    "Labels": {
      "license": "MIT",
      "name": "fedora",
      "vendor": "Fedora Project",
      "version": "32"
    }
  },
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:a4c0fa2b217d3fd63d51e55a6fd59432e543d499c0df2b1acd48fbe424f2ddd1"
    ]
  },
  "history": [
    {
      "created": "2020-04-29T06:48:16Z",
      "comment": "Created by Image Factory"
    }
  ]
}

# show unverifed image's digest

> skopeo inspect docker://registry.fedoraproject.org/fedora:latest | jq '.Digest'

Author: Fixed

"sha256:655721ff613ee766a4126cb5e0d5ae81598e1b0c3bcf7017c36c4d72cb092fe9"
```

Copying images
-
`skopeo` can copy container images between various storage mechanisms, including:
* Container registries

>   -  The Quay.io, Docker Hub, OpenShift, GCR, Artifactory ...

Reviewer: Should we include the RHEL and Fedora registries too?

Author: No, I think the list is enough.  Not sure Openshift should be there either, but will leave.  Want to maintain this as not Red Hat specific.

* Container Storage backends

  -  github.com/containers/storage (Backend for [Podman](https://podman.io), [CRI-O](https://cri-o.io), [Buildah](https://buildah.io) and friends)

  -  Docker daemon storage

* Local directories

* Local OCI-layout directories

```sh

> $ skopeo copy docker://quay.io/buildah/stable docker//registry.internal.company.com/buildah

Reviewer 2: ```suggestion
$ skopeo copy docker://quay.io/buildah/stable docker://registry.internal.company.com/buildah
```

Private registries with authentication
-

> When interacting with private registries, `skopeo` first looks for `--creds` (for `skopeo inspect|delete`) or `--src-creds|--dest-creds` (for `skopeo copy`) flags. If those aren't provided, it looks for the container authorization config file (usually located at `$XDG_RUNTIME_DIR/containers/auth.json`) to get the credentials needed to authenticate. The ultimate fallback, as other container engines do, is to provide an empty authentication when interacting with those registries.

Reviewer 2: Being completely silent about `~/.docker/config.json` is not helpful.

Author: I would think this is more a man page information rather then a README.md

Reviewer 2: To a large extent, most of this README duplicates the man page… especially this kind of detail should be only in there.

The way I think about it, the README can omit features and properties, but if we decide to say something, it should be _correct_.

Ultimately, the users don’t care and shouldn’t care about the file paths. What do you think about using just
> `skopeo` uses credentials from the `--creds` (for `skopeo inspect|delete`) or `--src-creds|--dest-creds` (for `skopeo copy`) flags, if set; otherwise it uses configuration set by `skopeo login`, `podman login`, `buildah login`, or `docker login`.

and dropping both the example and the final paragraph? We don’t want users to spend time `cat`ing `auth.json`, so this no longer serves much purpose in the top-level README.

Sure, it still refers to Docker — but it is so crowded in the list now :)

Author: SGTM

	}
}

# we can see I'm already authenticated via skopeo login so everything will be fine
$ skopeo inspect docker://myregistrydomain.com:5000/busybox
{"Tag":"latest","Digest":"sha256:473bb2189d7b913ed7187a33d11e743fdc2f88931122a44d91a301b64419f092","RepoTags":["latest"],"Comment":"","Created":"2016-01-15T18:06:41.282540103Z","ContainerConfig":{"Hostname":"aded96b43f48","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":null,"Cmd":["/bin/sh","-c","#(nop) CMD [\"sh\"]"],"Image":"9e77fef7a1c9f989988c06620dabc4020c607885b959a2cbd7c2283c91da3e33","Volumes":null,"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":null},"DockerVersion":"1.8.3","Author":"","Config":{"Hostname":"aded96b43f48","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":null,"Cmd":["sh"],"Image":"9e77fef7a1c9f989988c06620dabc4020c607885b959a2cbd7c2283c91da3e33","Volumes":null,"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":null},"Architecture":"amd64","Os":"linux"}

> # let's try now to fake a non existent authotization file

Reviewer: ```suggestion
# let's try now to fake a non-existent authorization file
```

$ cat /run/user/1000/containers/auth.json

> {}

Author: I don't know, I figured these two files were handled the same, so don't see a reason one would work and not the other.

> {}

Reviewer 2: I wasn’t paying enough attention — reading a `nil` map is OK in Go, it’s just writes that panic. (/me grumbles about questionable type/nil safety of that %$#@! language…)
